File: Olympiad/COCI/official/2007/contest3_solutions/npuzzle.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 3 - Task NPUZZLE
  Programming language C++
*/

#include <cstdio>

char a[4][5];
char b[4][5] = { "ABCD", "EFGH", "IJKL", "MNO#" };

int abs( int x ) { return x < 0 ? -x : x; }

int main( void ) {
   for( int r = 0; r < 4; ++r ) scanf( "%s", a[r] );

   int razbacanost = 0;

   for( int r1 = 0; r1 < 4; ++r1 )
      for( int s1 = 0; s1 < 4; ++s1 )
         for( int r2 = 0; r2 < 4; ++r2 )
            for( int s2 = 0; s2 < 4; ++s2 )
               if( a[r1][s1] == b[r2][s2] )
                  razbacanost += abs(r2-r1) + abs(s2-s1);

   printf( "%d\n", razbacanost );

   return 0;
}

File: Olympiad/COCI/official/2007/contest3_solutions/npuzzle.pas

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 3 - Task NPUZZLE
  Programming language Pascal
*/

program npuzzle;

var
   a : array[1..4,1..5] of char;
   b : array[1..4,1..5] of char;
   i, razbacanost : integer;
   r1, s1 : integer;
   r2, s2 : integer;

begin
   b[1] := 'ABCD';
   b[2] := 'EFGH';
   b[3] := 'IJKL';
   b[4] := 'MNO#';

   for i := 1 to 4 do readln( a[i] );

   razbacanost := 0;

   for r1 := 1 to 4 do
      for s1 := 1 to 4 do
         for r2 := 1 to 4 do
            for s2 := 1 to 4 do
               if a[r1,s1] = b[r2,s2] then
                  razbacanost := razbacanost + abs(r1-r2) + abs(s1-s2);

   writeln( razbacanost );
end.
